PagingWithNetworkSample/lib/src/main/java/com/android/example/paging/pagingwithnetwork/reddit/paging/SynchronousRemoteStates.kt

```diff
@@ -0,0 +1,141 @@
+        computeSynchronousRemoteStates(
+            sourceRefreshState = combinedLoadStates.source.refresh,
+            sourceState = combinedLoadStates.source.refresh,
+            remoteState = combinedLoadStates.mediator?.refresh,
+            synchronousRemoteState = refreshState,
+        ).also {
+            refresh = it.first
+            refreshState = it.second
+        }
+        computeSynchronousRemoteStates(
+            sourceRefreshState = combinedLoadStates.source.refresh,
+            sourceState = combinedLoadStates.source.prepend,
+            remoteState = combinedLoadStates.mediator?.prepend,
+            synchronousRemoteState = prependState,
+        ).also {
+            prepend = it.first
+            prependState = it.second
+        }
+        computeSynchronousRemoteStates(
+            sourceRefreshState = combinedLoadStates.source.refresh,
+            sourceState = combinedLoadStates.source.append,
+            remoteState = combinedLoadStates.mediator?.append,
+            synchronousRemoteState = appendState,
+        ).also {
+            append = it.first
+            appendState = it.second
+        }
+    }
+
+    private fun computeSynchronousRemoteStates(
+        sourceRefreshState: LoadState,
+        sourceState: LoadState,
+        remoteState: LoadState?,
+        synchronousRemoteState: SynchronousRemoteState,
+    ): Pair<LoadState, SynchronousRemoteState> {
+        if (remoteState == null) return sourceState to NOT_LOADING
+
+        return when (synchronousRemoteState) {
+            NOT_LOADING -> when (remoteState) {
+                is Loading -> Loading to REMOTE_STARTED
+                is Error -> remoteState to REMOTE_ERROR
+                else -> NotLoading(remoteState.endOfPaginationReached) to NOT_LOADING
+            }
+            REMOTE_STARTED -> when {
+                remoteState is Error -> remoteState to REMOTE_ERROR
+                sourceRefreshState is Loading -> Loading to SOURCE_LOADING
+                else -> Loading to REMOTE_STARTED
+            }
+            REMOTE_ERROR -> when (remoteState) {
+                is Error -> remoteState to REMOTE_ERROR
+                else -> Loading to REMOTE_STARTED
+            }
+            SOURCE_LOADING -> when {
+                sourceRefreshState is Error -> sourceRefreshState to SOURCE_ERROR
+                remoteState is Error -> remoteState to REMOTE_ERROR
+                sourceRefreshState is NotLoading -> {
+                    NotLoading(remoteState.endOfPaginationReached) to NOT_LOADING
+                }
+                else -> Loading to SOURCE_LOADING
+            }
+            SOURCE_ERROR -> when (sourceRefreshState) {
+                is Error -> sourceRefreshState to SOURCE_ERROR
+                else -> sourceRefreshState to SOURCE_LOADING
+            }
+        }
+    }
+}
+
+@OptIn(ExperimentalCoroutinesApi::class)
+fun Flow<CombinedLoadStates>.asSynchronousRemoteStates(): Flow<LoadStates> {
```

Comment:
add comments -

clarify when this can / can't be used, and why it's safe to use it in this sample (namely, behavior in room / remotemediator impl here)

I'm trying to remember - how does this work if remote tries to remote fetch, and remote sends the last page, with no next key? There won't be a local refresh, right?

should also explain what synchronousRemoteStates means - remote states, but deferred until it will synchronize with UI (e.g. remote loaded data already displayed).

Reply:
Room always invalidates even if no changes are posted so we are able to rely on that behavior here. After the last page, we shouldn't be making an "empty" check for endOfPaginationReached, we should just send that directly with the last page and invalidate if that makes sense? Adding those comments sgtm though.

Reply:
Done.